lib/opam_solve.ml

```diff
@@ -251,6 +254,84 @@ module Make_solver (Context : OPAM_MONOREPO_CONTEXT) :
+    let open Option.O in
+    let restrictions =
+      component |> Solver.Diagnostics.Component.notes
+      |> List.map ~f:(function
+           | Solver.Diagnostics.Note.UserRequested restriction ->
+               [ restriction ]
+           | Restricts (_other_role, _impl, restrictions) -> restrictions
+           | _ -> [])
+      |> List.flatten
+    in
+    let* version_restrictions =
+      match restrictions with
+      | [] -> None
+      | restrictions ->
+          restrictions
+          |> List.map ~f:(fun restriction ->
+                 let _, version_restriction = Solver.formula restriction in
+                 version_restriction)
+          |> Option.some
+    in
+    match version_restrictions with
+    | [] -> None
+    | init :: rest ->
+        List.fold_left
+          ~f:(fun formula restriction -> OpamFormula.And (formula, restriction))
+          ~init rest
+        |> Option.some
+
+  (* [true] if the rejected package would've satisified the version constraint if it
+     wouldn't have been a [Model_rejection] *)
+  let model_rejection_of_eligible_version version_restriction (model, reason) =
+    match reason with
+    | `Model_rejection _ -> (
+        match Solver.version model with
+        | None -> true
+        | Some rejected_package ->
+            let rejected_version = OpamPackage.version rejected_package in
+            let would_be_eligible_otherwise =
+              OpamFormula.check_version_formula version_restriction
+                rejected_version
+            in
+            would_be_eligible_otherwise)
+    | _ -> false
+
+  let unavailable_versions_due_to_constraints diagnostics =
+    let rolemap = Solver.diagnostics_rolemap diagnostics in
+    Pkg_map.fold
+      (fun pkg component unavailable ->
+        match Solver.Diagnostics.Component.selected_impl component with
+        | Some _ -> unavailable
+        | None ->
+            (* short-circuit skip of fold *)
+            let ( let* ) a f =
+              match a with Some a -> f a | None -> unavailable
+            in
+            let* pkg_name = Solver.package_name pkg in
+            let* version_restriction =
+              determine_version_restriction component
+            in
+            let rejects, _reason =
+              Solver.Diagnostics.Component.rejects component
+            in
+            (* check if any packages would have had matching versions if they weren't model-rejected *)
+            let model_rejections_would_match_version =
+              List.exists
+                ~f:(model_rejection_of_eligible_version version_restriction)
+                rejects
+            in
+            (* if it is unavailable, construct info on why *)
+            let* unavailable_pkg =
+              match model_rejections_would_match_version with
+              | false -> None
+              | true -> Some (pkg_name, version_restriction)
+            in
+            unavailable_pkg :: unavailable)
```

Comment:
The use of the option and the `let*` here is a bit confusing I think, especially since this is the returning block in a way.
What would you think about turning this into:
```suggestion
            match model_rejections_would_match_version with
            | false -> unavailable
            | true -> (pkg_name, version_restriction)::unavailable)
```